src/main/java/rx/observables/AsyncOnSubscribe.java

```diff
@@ -0,0 +1,510 @@
+            } else {
+                hasTerminated = true;
+                merger.onError(ex);
+                unsubscribe();
+            }
+        }
+
+        @Override
+        public void onCompleted() {
+            if (hasTerminated) {
+                throw new IllegalStateException("Terminal event already emitted.");
+            }
+            hasTerminated = true;
+            merger.onCompleted();
+        }
+
+        @Override
+        public void onError(Throwable e) {
+            if (hasTerminated) {
+                throw new IllegalStateException("Terminal event already emitted.");
+            }
+            hasTerminated = true;
+            merger.onError(e);
+        }
+
+        // This exists simply to check if the subscription has already been
+        // terminated before getting access to the subscription
+        private static Subscription SUBSCRIPTION_SENTINEL = new BooleanSubscription();
+
+        @Override
+        public void onNext(final Observable<? extends T> t) {
+            if (onNextCalled) {
+                throw new IllegalStateException("onNext called multiple times!");
+            }
+            onNextCalled = true;
+            if (hasTerminated)
+                return;
+            subscribeBufferToObservable(t);
+        }
+
+        private void subscribeBufferToObservable(final Observable<? extends T> t) {
+            BufferUntilSubscriber<T> buffer = BufferUntilSubscriber.<T> create();
+            final AtomicReference<Subscription> holder = new AtomicReference<Subscription>(null);
+            Subscription innerSubscription = t
+                .doOnTerminate(new Action0() {
+                    @Override
+                    public void call() {
+                        if (!holder.compareAndSet(null, SUBSCRIPTION_SENTINEL)) {
+                            Subscription h = holder.get();
+                            subscriptions.remove(h);
+                        }
+                    }})
+                .subscribe(buffer);
+
+            if (holder.compareAndSet(null, innerSubscription)) {
+                subscriptions.add(innerSubscription);
+            }
+            merger.onNext(buffer);
+        }
+    }
+
+    private static final class UnicastSubject<T> extends Observable<T>implements Observer<T> {
+        public static <T> UnicastSubject<T> create() {
+            return new UnicastSubject<T>(new State<T>());
+        }
+
+        private State<T> state;
+
+        protected UnicastSubject(final State<T> state) {
+            super(new OnSubscribe<T>() {
+                @Override
+                public void call(Subscriber<? super T> s) {
+                    if (state.subscriber != null) {
+                        s.onError(new IllegalStateException("There can be only one subscriber"));
+                    } else {
+                        state.subscriber = s;
+                    }
+                }
```

Comment:
I'm not sure if this class could get used by multiple concurrent subscribers (?) but if it does then I would expect to see an atomic compareAndSet here rather than just !=null.